### DISCUSSION TITLE
Scalability [graph performance] - Revisiting design of health on the graph

### DISCUSSION THREAD

**Discussion by jshaughn** [kiali/kiali#4507] [Roadmap]
This is related to the overall [Scalability Design Discussion](https://github.com/kiali/kiali/discussions/4080).

Graph Scalability is complicated, there are both usability and performance aspects.  The goal of this discussion is to see if we can improve performance by reworking the interaction of graph generation and health determination.  And then hopefully generate an issue to drive implementation.

Graph-gen performance deteriorates as the number of namespaces and/or the number of nodes increases.  There are a variety of factors in play, but in short each namespace is in essence its own graph, and those graphs are stitched together to form the final graph.  Furthermore, the number of nodes within each namespace increases the complexity.

In my opinion, the biggest gain for graph generation performance would be an improvement in the way we add health information.  Today we generate the graph and then, given the results, re-query for health information for the resulting graph components.  In practice this health determination can sometimes take as long as the graph generation itself.

In the current scalability discussions we have talked about maintaining a kiali “model” of the topology.  I think this could be very helpful, although I think I’d champion a “partial” model as opposed to what I think has been proposed in some of the talks.  I don’t currently think it should contain traffic metric data, but instead could be a more robust, proactively updated version of what we currently maintain in the cache.

One of the great strengths of Kiali today is it’s flexibility in querying real-time metrics for different durations, often updated by users while exploring their mesh.  We even allow querying past time periods in features like graph replay.  I don’t think we want to lose that ability, and I’m not sure we need to.  If nothing else, we don’t need to move from our current approach to a full “model” in one step.  I think we can start by making some improvements that could go a long way towards helping today’s graph performance.

### Idea: Separate Traffic Health and non-traffic Health

Traffic has history, and health can be determined using traffic from any time-period (although the health configuration would always be the current definition).  Non-traffic health: pod health, synchronization, validations,  etc, is calculated only against current state.

Traffic health calculation could be optimized by using the metric data already queried for the graph construction. Today, we use separate (but similar) queries. This alone should be a large performance gain while also relieving pressure on Prometheus.

The non-traffic health information should be available via the cached “model”, which may be slightly dated but is likely to be valid.  It would, I think, just be a simple assignment.

The node health assignments would all be made server-side, as part of the graph generation.  Each type of health could actually be made optional (and for graph replay, the non-traffic health could probably be skipped).

Thoughts?

**Comment by lucasponce** [comment]
> I think this could be very helpful, although I think I’d champion a “partial” model as opposed to what I think has been proposed in some of the talks.

I'm not sure if I understand the "partial" model idea, happy to hear more about it.

I think that a new abstraction is necessary to detach the implementation detail about how the topology model is built, as we are introducing some technical debts due the strong link with the base prometheus source.

With a new abstraction we may have a model that can keep use the current way to be calculated/fetched, but it may introduce new changes in a more robust transparent way:
- Use the source of nodes from non-prometheus source (registry).
- Use non Istio metrics.
- Better combine other signals (tracing and logs) to build the model.
- And as commented in the main discussion, to focus in the "query" aspects and how to extract that information.

I don't think that this abstraction should mean that we may loose current capabilities or features, but it would trigger a redesign to scale better and perhaps focus more in the "query" side (powerful filtering, push model with the UI, instead of current poll, etc).

I share the idea to move the health calculation to the model, instead to let the UI to query different endpoints and compose the final result as it's being calculated today.

This is a technical debt identified but I think it would deserve a top-down design approach, and that it will be linked with the main/global model discussions.

**Reply by jshaughn** [reply]
> I'm not sure if I understand the "partial" model idea, happy to hear more about it.

It's probably a good time to hold some team meetings and flesh out a few more details of what the "model" may look like and what it may provide, at least in a first milestone.

But what I'm saying is that I think it would be great for Kiali to have an internal model of the mesh(es) it is dealing with , in memory, designed to refresh periodically or as needed, and to be able to quickly supply information required by Kiali.  In this example, I can imagine it being able to quickly provide health information for graph nodes.  What I mean by "partial" is that it not be traffic-based health, but anything outside of traffic, whether it be validations, istio sync, pod health, etc.

In that way we can generate a graph using direct queries to Prom (or Proms), like today, and so we can maintain all of the flexibility we have today, but then decorate it server-side with all of the health information.  Today we generate the graph and then query back for health, which then performs its own Prom queries as well as the other health determinations, and it is not efficient.  I think we can go far with scalability with this partial, non-metric, model.  And then evaluate where to go from there.
